src/Bicep.Core/TypeSystem/TypeValidationVisitor.cs

```diff
@@ -0,0 +1,141 @@
+        {
+
+        }
+
+        public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
+        {
+            var assignedType = typeManager.GetTypeInfo(syntax);
+            
+            var diagnostics = TypeValidator.GetExpressionAssignmentDiagnostics(typeManager, syntax.Body, assignedType);
+
+            syntaxDiagnostics[syntax] = diagnostics.ToImmutableArray();
+        }
+
+        public override void VisitOutputDeclarationSyntax(OutputDeclarationSyntax syntax)
+        {
+            var diagnostics = GetOutputDeclarationDiagnostics(syntax);
+
+            syntaxDiagnostics[syntax] = diagnostics.ToImmutableArray();
+        }
+
+        private IEnumerable<Diagnostic> GetOutputDeclarationDiagnostics(OutputDeclarationSyntax syntax)
+        {
+            var assignedType = typeManager.GetTypeInfo(syntax);
+            var valueType = typeManager.GetTypeInfo(syntax.Value);
+
+            // this type is not a property in a symbol so the semantic error visitor won't collect the errors automatically
+            if (valueType is ErrorTypeSymbol)
+            {
+                return valueType.GetDiagnostics();
+            }
+
+            if (TypeValidator.AreTypesAssignable(valueType, assignedType) == false)
+            {
+                return DiagnosticBuilder.ForPosition(syntax.Value).OutputTypeMismatch(assignedType.Name, valueType.Name).AsEnumerable();
+            }
+
+            return Enumerable.Empty<Diagnostic>();
+        }
+
+        private IEnumerable<Diagnostic> ValidateDefaultValue(ParameterDefaultValueSyntax defaultValueSyntax, TypeSymbol assignedType)
+        {
+            // figure out type of the default value
+            var defaultValueType = typeManager.GetTypeInfo(defaultValueSyntax.DefaultValue);
+
+            // this type is not a property in a symbol so the semantic error visitor won't collect the errors automatically
+            if (defaultValueType is ErrorTypeSymbol)
+            {
+                return defaultValueType.GetDiagnostics();
+            }
+
+            if (TypeValidator.AreTypesAssignable(defaultValueType, assignedType) == false)
+            {
+                return DiagnosticBuilder.ForPosition(defaultValueSyntax.DefaultValue).ParameterTypeMismatch(assignedType.Name, defaultValueType.Name).AsEnumerable();
+            }
+
+            return Enumerable.Empty<Diagnostic>();
+        }
+
+        private IEnumerable<Diagnostic> ValidateIdentifierAccess(ParameterDeclarationSyntax syntax)
+        {
+            return SyntaxAggregator.Aggregate(syntax, new List<Diagnostic>(), (accumulated, current) =>
+                {
+                    if (current is VariableAccessSyntax)
+                    {
+                        var symbol = bindings[current];
+                        
+                        // excluded symbol kinds already generate errors - no need to duplicate
```

Comment:
> excluded symbol [](start = 26, length = 16)

Is this still true for Parameter and Output symbols?

Reply:
Yes, apart from params, which I believe are excluded for a different reason - because it is valid to reference a param in the default value as long as there are no cycles (which we check for elsewhere).

Reply:
I'll update the comment to elaborate a bit more on this.